### PR TITLE
#55 health check for external systems

### PR DESCRIPTION
**Pull request by TimoKramer**
Hi there, I am stuck with this one. Currently it starts the healtcheck but only checks once after a minute then does not check anymore. I removed the clojure.core.async/<!! from before the go-loop. There obviously is something missing. It should run continuously or am I missing something else? 

I tried using a thread but it does not quite help.

### CODE REVIEW

**Comment by TimoKramer** [comment]
@lispyclouds maybe you have another suggestion on what is missing here? and anything else that you see...

**Comment by lispyclouds** [comment]
Thanks a lot, I will try to have a look this weekend

**Comment by TimoKramer** [comment]
I introduced a mistake when I was starting from scratch...will fix this...

**Comment by TimoKramer** [comment]
so now it seems to work. @lispyclouds the heartbeat is currently in the health_check.clj under api, where do you think I should put the heartbeat functionality?

File: src/bob/api/health_check.clj [hunk 1]
```diff
@@ -15,28 +15,76 @@
...
+            [clj-time.periodic :as tp]
+            [bob.artifact.db :refer [get-artifact-stores]]
+            [bob.resource.db :refer [get-external-resources]]
             [bob.states :as states]))
 
 (sql/def-db-fns (io/resource "sql/health_checks.sql"))
 
+(defn ping-external-systems []
+  (let [external-systems (concat (get-artifact-stores states/db) (get-external-resources states/db))]
+    (map #(when (f/failed? (f/try* @(http/get (clojure.string/join "/" [(:url %) "ping"]) {:throw-exceptions false}))) (:name %))
```
Comment (lispyclouds) [inline src/bob/api/health_check.clj]: Remove the `{:throw-exceptions false}` to let it blow up and the `f/try*` will catch it for a proper failjure.

File: src/bob/api/health_check.clj [hunk 2]
```diff
@@ -15,28 +15,76 @@
...
-               postgres (when (f/failed? (f/try* (db-health-check states/db))) "Postgres")
-               failures (filter some? [docker postgres])]
+  (d/let-flow [docker   (when (f/failed? (f/try* (docker/ping states/docker-conn))) '("Docker"))
+               postgres (when (f/failed? (f/try* (db-health-check states/db))) '("Postgres"))
+               extsys   (when (nil? postgres) (ping-external-systems))]
+              (filter some? (concat docker postgres extsys))))
+
+(defn log-health-check []
+  (let [failures @(health-check)]
     (if (empty? failures)
```
Comment (lispyclouds) [inline src/bob/api/health_check.clj]: Use `when` here

File: src/bob/api/routes.clj [hunk 3]
```diff
@@ -179,7 +179,10 @@
       (rest/GET "/can-we-build-it" []
         :return schema/SimpleResponse
         :summary "Runs health checks for Bob."
-        (hc/health-check))
+          (let [failures @(hc/health-check)]
+            (if (empty? failures)
```
Comment (lispyclouds) [inline src/bob/api/routes.clj]: use `when` here

**Comment by lispyclouds** [comment]
health_check.clj is fine for this for now. Maybe we can just call the file `health.clj` for a more generic feel?

**Comment by lispyclouds** [comment]
Code looks good in general, try running on a local multi node bob cluster too and see if all works fine 😄

File: src/bob/api/routes.clj (hunk 3 above)
Comment (lispyclouds) [reply src/bob/api/routes.clj]: missed this. its okay then.

**Comment by lispyclouds** [comment]
This is good, merging.

**Pull request merged**